File: HealthApp/Tabs/Home/AddWorkout/Coordinator/AddWorkoutCoordinator.m
//
//  AddWorkoutCoordinator.m
//  HealthApp
//
//  Created by Christopher Ray on 3/27/21.
//

#import "AddWorkoutCoordinator.h"
#import "WorkoutViewController.h"
#import "AddWorkoutUpdateMaxesViewController.h"
#import "HomeTabCoordinator.h"
#import "AppDelegate.h"
#include "AppUserData.h"
#import "PersistenceService.h"
#import "WeeklyData+CoreDataClass.h"

void updateStoredData(AddWorkoutViewModel *model) {
    int duration = (int) ((model->stopTime - model->startTime) / 60.0);
    WeeklyData *data = persistenceService_getWeeklyDataForThisWeek();
    if (!(duration >= 15 && data)) return;

    switch (model->workout->type) {
        case WorkoutTypeSE:
            data.timeSE += duration;
            break;
        case WorkoutTypeHIC:
            data.timeHIC += duration;
            break;
        case WorkoutTypeStrength:
            data.timeStrength += duration;
            break;
        default:
            data.timeEndurance += duration;
            break;
    }
    data.totalWorkouts += 1;
    persistenceService_saveContext();
}

void didFinishAddingWorkout(AddWorkoutCoordinator *this, UIViewController *presenter, int totalCompletedWorkouts) {
    if (presenter) {
        [presenter dismissViewControllerAnimated:true completion:nil];
        AppDelegate *delegate = (AppDelegate *) UIApplication.sharedApplication.delegate;
        if (delegate) appCoordinator_updateMaxWeights(&delegate->coordinator);
    }
    homeCoordinator_didFinishAddingWorkout(this->parent, totalCompletedWorkouts);
}

void addWorkoutCoordinator_start(AddWorkoutCoordinator *this) {
    UIViewController *vc = [[WorkoutViewController alloc] initWithDelegate:this];
    [this->navigationController pushViewController:vc animated:true];
    [vc release];
}

void addWorkoutCoordinator_free(AddWorkoutCoordinator *this) {
    array_free(exGroup, this->viewModel.workout->activities);
    CFRelease(this->viewModel.workout->title);
    free(this->viewModel.workout);
    free(this);
}

void addWorkoutCoordinator_stoppedWorkout(AddWorkoutCoordinator *this) {
    updateStoredData(&this->viewModel);
    didFinishAddingWorkout(this, nil, 0);
}

void addWorkoutCoordinator_completedWorkout(AddWorkoutCoordinator *this, UIViewController *presenter,
                                            bool showModalIfRequired) {
    CFStringRef title = this->viewModel.workout->title;
    if (showModalIfRequired && CFStringCompareWithOptions(title, CFSTR("test day"),
                                                          CFRangeMake(0, CFStringGetLength(title)),
                                                          kCFCompareCaseInsensitive) == kCFCompareEqualTo) {
        UIViewController *modal = [[AddWorkoutUpdateMaxesViewController alloc] initWithDelegate:this];
        UINavigationController *container = [[UINavigationController alloc] initWithRootViewController:modal];
        [this->navigationController presentViewController:container animated:true completion:nil];
        [container release];
        [modal release];
        return;
    }

    updateStoredData(&this->viewModel);
    const signed char day = this->viewModel.workout->day;
    if (day >= 0) {
        unsigned char totalCompleted = appUserData_addCompletedWorkout((unsigned char) day);
        didFinishAddingWorkout(this, presenter, totalCompleted);
        return;
    }
    didFinishAddingWorkout(this, presenter, 0);
}

void addWorkoutCoordinator_finishedAddingNewWeights(AddWorkoutCoordinator *this, UIViewController *presenter,
                                                    short *weights) {
    WeeklyData *data = persistenceService_getWeeklyDataForThisWeek();
    if (data) {
        data.bestSquat = weights[0];
        data.bestPullup = weights[1];
        data.bestBench = weights[2];
        data.bestDeadlift = weights[3];
        persistenceService_saveContext();
    }
    appUserData_updateWeightMaxes(weights);
    addWorkoutCoordinator_completedWorkout(this, presenter, 0);
}

void addWorkoutCoordinator_stopWorkoutFromBackButtonPress(AddWorkoutCoordinator *this) {
    if (this->viewModel.startTime) {
        this->viewModel.stopTime = CFAbsoluteTimeGetCurrent();
        updateStoredData(&this->viewModel);
    }
    addWorkoutCoordinator_free(this);
}

File: HealthApp/Tabs/Home/AddWorkout/Coordinator/AddWorkoutCoordinator.c
#include "AddWorkoutCoordinator.h"
#include "AppCoordinator.h"
#include "AppUserData.h"
#include "PersistenceService.h"
#include "ViewControllerHelpers.h"
#include "WorkoutVC.h"
#include "UpdateMaxesModal.h"

extern void homeCoordinator_didFinishAddingWorkout(void *parent, int totalCompleted, bool popStack);

static bool checkEnduranceDuration(Workout *w) {
    if (w->type != WorkoutEndurance) return false;
    int planDuration = w->activities->arr[0].exercises->arr[0].reps / 60;
    return w->duration >= planDuration;
}

static void setDuration(Workout *w) {
    w->duration = ((int16_t) ((time(NULL) - w->startTime) / 60.f)) + 1;
#if TARGET_OS_SIMULATOR
    w->duration *= 10;
#endif
}

static void updateStoredDataAndCleanup(AddWorkoutCoordinator *this, short *lifts) {
    unsigned char type = this->workout->type;
    int16_t duration = this->workout->duration;

    CFRelease(this->workout->title);
    array_free(circuit, this->workout->activities);
    free(this->workout);
    free(this);

    if (duration < 15) return;

    runInBackground((^{
        id data = persistenceService_getCurrentWeek();
        int16_t newDuration = duration + weekData_getWorkoutTimeForType(data, type);
        weekData_setWorkoutTimeForType(data, type, newDuration);
        int16_t totalWorkouts = weekData_getTotalWorkouts(data);
        totalWorkouts += 1;
        weekData_setTotalWorkouts(data, totalWorkouts);
        if (lifts) {
            weekData_setLiftingMaxArray(data, lifts);
            free(lifts);
        }
        persistenceService_saveContext();
    }));
}

static void handleFinishedWorkout(AddWorkoutCoordinator *this,
                                  bool dismissVC, bool goBack, short *lifts) {
    const signed char day = this->workout->day;
    unsigned char totalCompleted = 0;
    bool longEnough = this->workout->duration >= 15;
    short *liftsCopy = NULL;
    if (longEnough) {
        if (day >= 0)
            totalCompleted = appUserData_addCompletedWorkout(day);
        if (lifts) {
            liftsCopy = malloc(sizeof(short) << 2);
            memcpy(liftsCopy, lifts, sizeof(short) << 2);
        }
    }

    id navVC = this->navVC;
    void *parent = this->parent;
    updateStoredDataAndCleanup(this, lifts);
    if (dismissVC) {
        dismissPresentedVC(navVC, ^{
            appCoordinator_updateMaxWeights(liftsCopy);
            homeCoordinator_didFinishAddingWorkout(parent, totalCompleted, true);
        });
    } else {
        if (liftsCopy)
            free(liftsCopy);
        homeCoordinator_didFinishAddingWorkout(parent, totalCompleted, goBack);
    }
}

void addWorkoutCoordinator_start(AddWorkoutCoordinator *this) {
    id vc = workoutVC_init(this);
    ((void(*)(id,SEL,id,bool))objc_msgSend)(this->navVC,
                                            sel_getUid("pushViewController:animated:"), vc, true);
    releaseObj(vc);
}

void addWorkoutCoordinator_stoppedWorkout(AddWorkoutCoordinator *this) {
    setDuration(this->workout);
    void *parent = this->parent;
    if (checkEnduranceDuration(this->workout)) {
        handleFinishedWorkout(this, false, true, NULL);
    } else {
        updateStoredDataAndCleanup(this, NULL);
        homeCoordinator_didFinishAddingWorkout(parent, 0, true);
    }
}

void addWorkoutCoordinator_completedWorkout(AddWorkoutCoordinator *this, bool dismissVC,
                                            bool showModalIfRequired, short *lifts) {
    CFStringRef title = this->workout->title;
    if (showModalIfRequired)
        setDuration(this->workout);
    if (showModalIfRequired &&
        CFStringCompareWithOptions(title, localize(CFSTR("workoutTitleTestDay")),
                                   CFRangeMake(0, CFStringGetLength(title)),
                                   kCFCompareCaseInsensitive) == kCFCompareEqualTo) {
        presentModalVC(this->navVC, updateMaxesVC_init(this));
    } else {
        handleFinishedWorkout(this, dismissVC, true, lifts);
    }
}

void addWorkoutCoordinator_stopWorkoutFromBackButtonPress(AddWorkoutCoordinator *this) {
    if (this->workout->startTime) {
        setDuration(this->workout);
        if (checkEnduranceDuration(this->workout)) {
            handleFinishedWorkout(this, false, false, NULL);
            return;
        }
    }
    updateStoredDataAndCleanup(this, NULL);
}
